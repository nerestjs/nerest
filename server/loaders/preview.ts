import path from 'path';
import { existsSync } from 'fs';
import fs from 'fs/promises';

export type PreviewParts = {
  head: string;
};

export async function loadPreviewParts(root: string): Promise<PreviewParts> {
  let head = '';
  const headPath = path.join(root, 'nerest/preview-head.html');
  if (existsSync(headPath)) {
    head = await fs.readFile(headPath, 'utf-8');
  }

  return { head };
}
