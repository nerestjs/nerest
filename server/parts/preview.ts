import type { PreviewParts } from '../loaders/preview.js';

// Renders the preview page available by convention at /api/{name}/examples/{example}
export function renderPreviewPage(
  html: string,
  assets: string[],
  parts: PreviewParts
) {
  const { scripts, styles } = mapAssets(assets);
  return `
  <!DOCTYPE html>
  <html>
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    ${styles.join('\n')}
    ${parts.head}
  </head>
  <body>
    ${html}
    ${scripts.join('\n')}
  </body>
  </html>
  `;
}

function mapAssets(assets: string[]) {
  const scripts = assets
    .filter((src) => src.endsWith('.js'))
    .map((src) => `<script type="module" src="${src}"></script>`);

  const styles = assets
    .filter((src) => src.endsWith('.css'))
    .map((src) => `<link rel="stylesheet" href="${src}">`);

  return { scripts, styles };
}
