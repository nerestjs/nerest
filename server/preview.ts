// Renders the preview page available by convention at /api/{name}/examples/{example}
export function renderPreviewPage(html: string, assets: string[]) {
  const { scripts, styles } = mapAssets(assets);
  return `
  <html>
  <head>
    ${styles.join('\n')}
  </head>
  <body>
    ${html}
    ${scripts.join('\n')}
  </body>
  </html>
  `;
}

function mapAssets(assets: string[]) {
  // TODO: script type="module" is not supported by older browsers
  // but vite doesn't provide `nomodule` fallback by default
  // see @vitejs/plugin-legacy
  const scripts = assets
    .filter((src) => src.endsWith('.js'))
    .map((src) => `<script type="module" src="${src}"></script>`);

  const styles = assets
    .filter((src) => src.endsWith('.css'))
    .map((src) => `<link rel="stylesheet" href="${src}">`);

  return { scripts, styles };
}
