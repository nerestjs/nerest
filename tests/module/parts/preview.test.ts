import { describe, it, expect } from 'vitest';
import { renderPreviewPage } from '../../../server/parts/preview.js';

describe('renderPreviewPage', () => {
  it('should render a complete HTML page with provided content', () => {
    const result = renderPreviewPage(
      '<div>Test content</div>',
      ['/assets/main.js', '/assets/style.css', '/assets/extra.js'],
      {
        head: '<title>Test Page</title>',
      }
    );

    expect(result).toMatchSnapshot();
  });

  it('should order CSS before JS in the document', () => {
    const result = renderPreviewPage(
      '<div>Content</div>',
      ['/assets/script.js', '/assets/style.css'],
      {
        head: '',
      }
    );

    // Convert to single line to make index comparison easier
    const singleLineResult = result.replace(/\n\s*/g, '');

    const cssIndex = singleLineResult.indexOf('/assets/style.css');
    const jsIndex = singleLineResult.indexOf('/assets/script.js');

    expect(cssIndex).toBeLessThan(jsIndex);
  });
});
