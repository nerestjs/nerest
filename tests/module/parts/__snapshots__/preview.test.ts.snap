// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renderPreviewPage > should render a complete HTML page with provided content 1`] = `
"
  <!DOCTYPE html>
  <html>
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" href="/assets/style.css">
    <title>Test Page</title>
  </head>
  <body>
    <div>Test content</div>
    <script type="module" src="/assets/main.js"></script>
<script type="module" src="/assets/extra.js"></script>
  </body>
  </html>
  "
`;
